org.jboss.tools.vscode.java/src/org/jboss/tools/vscode/java/internal/handlers/CodeActionHandler.java

```diff
@@ -0,0 +1,148 @@
+		return $;
+	}
+
+
+	private boolean hasCodeAssist(Diagnostic diagnostic){
+
+		final int problemId= getProblemId(diagnostic);
+		switch(problemId){
+		case IProblem.UnterminatedString:
+		case IProblem.UnusedImport:
+		case IProblem.DuplicateImport:
+		case IProblem.CannotImportPackage:
+		case IProblem.ConflictingImport:
+		case IProblem.ImportNotFound:
+			return true;
+		default:
+			return false;
+		}
+	}
+
+	private List<Command> getCommandForDiagnostic(ICompilationUnit unit, String uri,Diagnostic diagnostic){
+		final int problemId = getProblemId(diagnostic);
+		switch(problemId){
+		case IProblem.UnterminatedString:
+			String label = "Insert missing quote";
+			WorkspaceEdit edit = new WorkspaceEdit();
+			try {
+				int offset = DiagnosticsHelper.getEndOffset(unit, diagnostic);
+				int start = DiagnosticsHelper.getStartOffset(unit, diagnostic);
+				int pos = moveBack(offset, start, "\n\r", unit);
+				TextEdit te = new TextEdit(JDTUtils.toRange(unit, pos, 0), "\"");
+				edit.getChanges().put(uri, Arrays.asList(te));
+				return Arrays.asList(new Command(label, COMMAND_ID_APPLY_EDIT, Arrays.asList(edit)));
+			} catch (JavaModelException e) {
+				//return null;
+			}
+		case IProblem.UnusedImport:
+		case IProblem.DuplicateImport:
+		case IProblem.CannotImportPackage:
+		case IProblem.ConflictingImport:
+		case IProblem.ImportNotFound:
+			//TODO: Add Organize imports command when/if we support it
+			int start = DiagnosticsHelper.getStartOffset(unit, diagnostic);
+			int end = DiagnosticsHelper.getEndOffset(unit, diagnostic);
+			org.eclipse.text.edits.TextEdit te = UnusedCodeCorrections.createUnusedImportTextEdit(getASTRoot(unit), start, end-start);
+			return Arrays.asList(textEditToCommand(unit, uri, "Remove unused import", te));
+		default:
+			return null;
+		}
+	}
+
+	private int getProblemId(Diagnostic diagnostic){
+		int $=0;
+		try{
+			$ = Integer.parseInt(diagnostic.getCode());
+		}catch(NumberFormatException e){
+			// return 0
+		}
+		return $;
+	}
+
+
+
+	private static Command textEditToCommand(ICompilationUnit unit, String uri, String label, org.eclipse.text.edits.TextEdit textEdit){
+		TextEditConverter converter = new TextEditConverter(unit, textEdit);
+		List<TextEdit> edits = converter.convert();
+		WorkspaceEdit $ = new WorkspaceEdit();
+		$.getChanges().put(uri, edits);
+		return new Command(label, COMMAND_ID_APPLY_EDIT, Arrays.asList($));
+	}
+
+	private static CompilationUnit getASTRoot(ICompilationUnit unit){
+		return SharedASTProvider.getInstance().getAST(unit, new NullProgressMonitor());
+	}
+
+	private static int moveBack(int offset, int start, String ignoreCharacters, ICompilationUnit cu) {
+		try {
+			IBuffer buf= cu.getBuffer();
+			while (offset >= start) {
+				if (ignoreCharacters.indexOf(buf.getChar(offset - 1)) == -1) {
+					return offset;
+				}
+				offset--;
+			}
+		} catch(JavaModelException e) {
+			// use start
```

Comment:
shouldn't we ignore the whole change in case an exception occurred (by propagating this one or returning -1)?